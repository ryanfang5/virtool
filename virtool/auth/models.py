from sqlalchemy import Column, Enum, String

from virtool.pg.base import Base
from virtool.pg.utils import SQLEnum


class ResourceType(str, SQLEnum):
    """
    Types of resources a permission can apply to (e.g. 'app', 'sample', 'group').

    """

    app = "app"
    group = "group"


class Action(str, SQLEnum):
    """
    Conserved actions that can be performed on a resource and
    are controlled by permissions.
    """

    create = "create"
    update = "update"
    delete = "delete"
    modify = "modify"
    remove = "remove"


class SQLPermission(Base):
    """
    SQL model to store permissions

    """

    __tablename__ = "permissions"

    id = Column(String, primary_key=True)
    name = Column(String)
    description = Column(String)
    resource_type = Column(Enum(ResourceType))
    action = Column(Enum(Action))
