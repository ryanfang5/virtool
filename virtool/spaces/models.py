from sqlalchemy import (
    Column,
    DateTime,
    Integer,
    String,
)

from virtool.pg.base import Base


class SpaceModel(Base):
    __tablename__ = "space"

    id = Column(Integer, primary_key=True)
    name = Column(String, unique=True)
    description = Column(String)
    created_at = Column(DateTime, nullable=False)
    updated_at = Column(DateTime, nullable=False)
    created_by = Column(String, nullable=False)

