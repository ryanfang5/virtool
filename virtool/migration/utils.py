from pathlib import Path


def get_revisions_path() -> Path:
    """
    Get the hardcoded path to data revision files.

    :return: the revisions path

    """
    return Path(__file__).parent.parent.parent / "assets/revisions"
