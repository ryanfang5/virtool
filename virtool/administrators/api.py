from typing import Union

from aiohttp.web_exceptions import HTTPNoContent, HTTPBadRequest
from aiohttp_pydantic import PydanticView
from aiohttp_pydantic.oas.typing import r201, r200, r204, r404

from virtool_core.models.roles import AdministratorRole

from virtool.administrators.oas import (
    CreateAdministratorRequest,
    UpdateAdministratorRequest,
    ListAdministratorResponse,
    GetAdministratorResponse,
    CreateAdministratorResponse,
    UpdateAdministratorResponse,
)
from virtool.api.response import NotFound, json_response
from virtool.data.errors import ResourceNotFoundError
from virtool.data.utils import get_data_from_req
from virtool.http.policy import policy, AdministratorRoutePolicy
from virtool.http.routes import Routes

routes = Routes()


@routes.view("/administrators")
class AdministratorsView(PydanticView):
    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
    async def get(self) -> r200[ListAdministratorResponse]:
        """
        List administrators.

        Lists all administrators.

        Status Codes:
            200: Successful operation
        """
        return json_response(
            await get_data_from_req(self.request).administrators.find()
        )

    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
    async def post(
        self, data: CreateAdministratorRequest
    ) -> Union[r201[CreateAdministratorResponse], r404]:
        """
        Add an administrator.

        Adds a user as an administrator using their user id.

        Status Codes:
            201: Successful operation
            400: User not found
        """

        try:
            administrator = await get_data_from_req(self.request).administrators.create(
                data
            )
        except ResourceNotFoundError as err:
            if "User not found" in str(err):
                raise HTTPBadRequest(text=str(err))

        return json_response(administrator, status=201)


@routes.view("/administrators/{user_id}")
class AdministratorView(PydanticView):
    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
    async def get(self, user_id: str, /) -> Union[r200[GetAdministratorResponse], r404]:
        """
        Get an administrator.

        Fetches the complete representation of an administrator.

        Status Codes:
            200: Successful operation
            404: User not found
        """

        try:
            administrator = await get_data_from_req(self.request).administrators.get(
                user_id
            )
        except ResourceNotFoundError:
            raise NotFound()

        return json_response(administrator)

    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
    async def patch(
        self, user_id: str, /, data: UpdateAdministratorRequest
    ) -> Union[r201[UpdateAdministratorResponse], r404]:
        """
        Update administrator's role.

        Updates administrator data for a specified user.

        Administrators cannot modify their own role.

        Status Codes:
            200: Successful operation
            400: Cannot modify own role
            404: User not found
        """
        if user_id == self.request["client"].user_id:
            raise HTTPBadRequest(text="Administrators cannot modify their own role.")

        try:
            administrator = await get_data_from_req(self.request).administrators.update(
                user_id, data
            )
        except ResourceNotFoundError:
            raise NotFound()

        return json_response(administrator)

    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
    async def delete(self, user_id: str, /) -> Union[r204, r404]:
        """
        Remove an administrator.

        Removes a user as an administrator.

        Status Codes:
            204: Successful operation
            404: User not found
        """

        if user_id == self.request["client"].user_id:
            raise HTTPBadRequest(
                text="Users cannot modify their own administrative status"
            )

        try:
            await get_data_from_req(self.request).administrators.delete(user_id)
        except ResourceNotFoundError:
            raise NotFound()

        raise HTTPNoContent
