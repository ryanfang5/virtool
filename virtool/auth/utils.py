import sys
from enum import Enum
from logging import getLogger

import openfga_sdk
from aiohttp import ClientConnectorError
from openfga_sdk import (
    CreateStoreRequest,
    WriteAuthorizationModelRequest,
    TypeDefinition,
    Userset,
    OpenFgaApi,
    WriteRequest,
    TupleKey,
    ApiException,
    TupleKeys,
)
from openfga_sdk.api import open_fga_api

from virtool.auth.permissions import (
    AppPermission,
)

logger = getLogger("openfga")


async def connect_openfga(openfga_host: str, openfga_scheme: str):
    """
    Connects to an OpenFGA server and configures the store id.
    Returns the application client instance.
    """
    configuration = openfga_sdk.Configuration(
        api_scheme=openfga_scheme, api_host=openfga_host
    )

    logger.info("Connecting to OpenFGA")

    try:
        api_client = openfga_sdk.ApiClient(configuration)

        api_instance = open_fga_api.OpenFgaApi(api_client)

        configuration.store_id = await get_or_create_store(api_instance)

        await write_auth_model(api_instance)

    except ClientConnectorError:
        logger.critical("Could not connect")
        sys.exit(1)

    return api_instance


async def get_or_create_store(api_instance: OpenFgaApi):
    """
    Get the OpenFGA Store or create one if it does not exist.

    :return: the store id
    """
    response = await api_instance.list_stores()

    logger.info("Connected to OpenFGA")

    if response.stores:
        logger.info("Found existing store")
        return response.stores[0].id

    body = CreateStoreRequest(
        name="Virtool",
    )
    logger.info("Creating store")
    response = await api_instance.create_store(body)
    return response.id


async def write_auth_model(api_instance: OpenFgaApi):
    """
    Write the authorization model for the OpenFGA Store if it does not exist.
    """
    response = await api_instance.read_authorization_models()

    app_definition = TypeDefinition(
        type="app",
        relations={permission.name: Userset(this={}) for permission in AppPermission},
    )

    group_definition = TypeDefinition(
        type="group",
        relations={"member": Userset(this={})},
    )

    type_definitions = WriteAuthorizationModelRequest(
        type_definitions=[app_definition, group_definition],
    )

    if (
        response.authorization_models
        and response.authorization_models[0].type_definitions
        == type_definitions.type_definitions
    ):
        return

    await api_instance.write_authorization_model(type_definitions)


async def write_tuple(
    api_instance, user_type, user_id, relations, object_type, object_name
):
    """
    Write a relationship tuple in OpenFGA.
    """

    tuple_list = [
        TupleKey(
            user=f"{user_type}:{user_id}",
            relation=relation.name if isinstance(relation, Enum) else relation,
            object=f"{object_type}:{object_name}",
        )
        for relation in relations
    ]

    body = WriteRequest(writes=TupleKeys(tuple_keys=tuple_list))

    try:
        await api_instance.write(body)
    except ApiException:
        pass
