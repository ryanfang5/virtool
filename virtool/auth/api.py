from typing import Optional

from aiohttp_pydantic import PydanticView
from aiohttp_pydantic.oas.typing import r200

from virtool.api.response import json_response
from virtool.auth.models import ResourceType
from virtool.auth.oas import ListPermissionsResponse
from virtool.data.utils import get_data_from_req
from virtool.http.routes import Routes

routes = Routes()


@routes.view("/source/permissions")
class SourceView(PydanticView):
    async def get(
        self, resource_type: Optional[ResourceType] = None
    ) -> r200[ListPermissionsResponse]:
        """
        List permissions.

        Lists all Virtool permissions.

        The list can be filtered by resource type using the
        `resource_type` query parameter.

        Status Codes:
            200: Successful operation
        """

        return json_response(
            await get_data_from_req(self.request).auth.find(resource_type)
        )
