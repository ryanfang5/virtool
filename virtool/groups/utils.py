from __future__ import annotations

from copy import deepcopy
from typing import List, Dict, Tuple

import virtool.users.utils
from virtool.authorization.permissions import SpacePermission
from virtool.authorization.relationships import GroupPermission
from virtool.users.utils import Permission


def merge_group_permissions(groups: List[dict]) -> Dict[str, bool]:
    """
    Return a :class:`dict` of permissions that will be
    inherited by a user belonging to all the passed ``groups``.

    :param groups: a list of group documents.
    :return: a dict keyed by permission names with boolean values
    indicating the state of the permission

    """
    permission_dict = virtool.users.utils.generate_base_permissions()

    for p in Permission:
        for group in groups:
            try:
                if group["permissions"][p.value]:
                    permission_dict[p.value] = True
            except KeyError:
                permission_dict[p.value] = False

    return permission_dict


def adapt_legacy_permissions_dict(permissions_dict: dict[str, bool]):
    adapted = deepcopy(permissions_dict)

    adapted.pop("remove_job", None)

    adapted["create_reference"] = adapted.pop("create_ref")
    adapted["delete_upload"] = adapted.pop("remove_file")
    adapted["create_upload"] = adapted.pop("upload_file")

    return adapted


def convert_permissions_dict_to_relationships(
    group_id: str,
    permissions_dict: dict[str, bool],
) -> Tuple[List[GroupPermission], List[GroupPermission]]:
    adapted_permissions_dict = adapt_legacy_permissions_dict(permissions_dict)

    adds = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if value
    ]

    removes = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if not value
    ]

    return adds, removes
