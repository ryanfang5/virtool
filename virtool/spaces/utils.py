import asyncio
from datetime import datetime
from enum import EnumMeta
from logging import getLogger
from typing import List, Union, Optional

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import (
    SpaceRole,
    SpaceLabelRole,
    SpaceProjectRole,
    SpaceReferenceRole,
    SpaceSampleRole,
    SpaceSubtractionRole,
    SpaceUploadRole,
)
from virtool_core.models.user import UserNested
from virtool_core.utils import document_enum

from virtool.authorization.client import AuthorizationClient
from virtool.authorization.relationships import UserRoleAssignment
from virtool.mongo.core import Mongo

logger = getLogger(__name__)


class SpaceMember(UserNested):
    role: SpaceRole
    label_role: Optional[SpaceLabelRole]
    project_role: Optional[SpaceProjectRole]
    reference_role: Optional[SpaceReferenceRole]
    sample_role: Optional[SpaceSampleRole]
    subtraction_role: Optional[SpaceSubtractionRole]
    upload_role: Optional[SpaceUploadRole]


class SpaceNested(BaseModel):
    id: int
    name: str


class SpaceMinimal(SpaceNested):
    description: str


class Space(SpaceMinimal):
    created_at: datetime
    updated_at: datetime
    created_by: str
    members: List[SpaceMember]


class MemberSearchResult(BaseModel):
    items: List[SpaceMember]
    available_roles: List[dict]


AVAILABLE_ROLES = [
    {"id": role, "name": role.capitalize(), "description": role.__doc__}
    for enum in [
        SpaceRole,
        SpaceLabelRole,
        SpaceProjectRole,
        SpaceReferenceRole,
        SpaceSampleRole,
        SpaceSubtractionRole,
        SpaceUploadRole,
    ]
    for role in document_enum(enum)
]


async def remove_user_roles(
    authorization_client: AuthorizationClient,
    user_id: Union[str, int],
    space_id: int,
    enums: List[EnumMeta],
):

    await asyncio.gather(
        *[
            authorization_client.remove(
                UserRoleAssignment(user_id, space_id, enum_type(role))
            )
            for enum_type in enums
            for role in enum_type
        ]
    )


def format_user(user: dict, role_list: List):
    user = {
        **user,
        "label_role": None,
        "project_role": None,
        "reference_role": None,
        "sample_role": None,
        "subtraction_role": None,
        "upload_role": None,
    }

    for role in role_list:
        if role in iter(SpaceRole):
            user["role"] = SpaceRole(role)

        if role in iter(SpaceLabelRole):
            user["label_role"] = SpaceLabelRole(role)

        if role in iter(SpaceProjectRole):
            user["project_role"] = SpaceProjectRole(role)

        if role in iter(SpaceReferenceRole):
            user["reference_role"] = SpaceReferenceRole(role)

        if role in iter(SpaceSampleRole):
            user["sample_role"] = SpaceSampleRole(role)

        if role in iter(SpaceSubtractionRole):
            user["subtraction_role"] = SpaceSubtractionRole(role)

        if role in iter(SpaceUploadRole):
            user["upload_role"] = SpaceUploadRole(role)

    return SpaceMember(**user)


async def format_users(
    authorization_client: AuthorizationClient, mongo: Mongo, space_id: int
) -> List[SpaceMember]:

    member_ids = await authorization_client.list_space_users(space_id)

    users = await mongo.users.find(
        {"_id": {"$in": [member[0] for member in member_ids]}}
    ).to_list(None)

    if len(users) != len(member_ids):
        logger.warning("Missing users.")

    member_ids = {member[0]: member[1] for member in member_ids}

    return [format_user(user, member_ids[user["_id"]]) for user in users]
