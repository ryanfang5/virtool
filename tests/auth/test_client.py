import openfga_sdk
import pytest
from openfga_sdk.api import open_fga_api
from virtool_core.models.enums import Permission

from virtool.auth.utils import write_tuple


@pytest.fixture()
async def delete_store():
    """
    Delete the store.
    """

    configuration = openfga_sdk.Configuration(
        api_scheme="http", api_host="localhost:8080"
    )
    api_client = open_fga_api.ApiClient(configuration)
    api_instance = open_fga_api.OpenFgaApi(api_client)

    response = await api_instance.list_stores()

    for store in response.stores:
        configuration.store_id = store.id
        await api_instance.delete_store()


@pytest.mark.parametrize("db", ["mongo", "open_fga"])
@pytest.mark.parametrize("has_permission", [True, False])
async def test_check(delete_store, db, has_permission, spawn_client):
    client = await spawn_client(authorize=True, permissions=[Permission.cancel_job])
    abs_client = client.app["auth"]

    if has_permission:
        permission = Permission.cancel_job
    else:
        permission = Permission.modify_subtraction

    if db == "mongo":
        response = await abs_client.check("test", permission, "app", "virtool")

        # assert await check_in_mongo(abs_client.mongo, "test", permission) is has_permission

    else:
        await write_tuple(
            abs_client.open_fga,
            "user",
            "ryanf",
            Permission.cancel_job,
            "app",
            "virtool",
        )

        response = await abs_client.check("ryanf", permission, "app", "virtool")

        # assert await check_in_open_fga(abs_client.open_fga, "ryanf", permission, "app", "virtool") is has_permission

    assert response is has_permission


@pytest.mark.parametrize("db", ["mongo", "open_fga"])
async def test_list(delete_store, db, spawn_client, snapshot):
    client = await spawn_client(authorize=True, permissions=[Permission.cancel_job])

    abs_client = client.app["auth"]

    if db == "mongo":
        response = await abs_client.list_permissions("test", "app", "virtool")
    else:
        await write_tuple(
            abs_client.open_fga, "user", "ryanf", "member", "group", "sidney"
        )

        await write_tuple(
            abs_client.open_fga,
            "group",
            "sidney#member",
            "cancel_job",
            "app",
            "virtool",
        )

        await write_tuple(
            abs_client.open_fga,
            "group",
            "sidney#member",
            "create_ref",
            "app",
            "virtool",
        )

        response = await abs_client.list_permissions("ryanf", "app", "virtool")

    assert response == snapshot


@pytest.mark.parametrize("db", ["mongo", "open_fga"])
async def test_add_group_permissions(delete_store, setup_update_group, db, snapshot):
    client, group = setup_update_group

    abs_client = client.app["auth"]

    if db == "mongo":

        await abs_client.add_group_permissions(
            group.id,
            [Permission.cancel_job, Permission.modify_subtraction],
            "app",
            "virtool",
        )

        assert (
                await client.db.users.find({}, ["groups", "permissions"]).to_list(None)
                == snapshot
        )

    else:

        await write_tuple(
            abs_client.open_fga, "user", "ryanf", "member", "group", "sidney"
        )

        await abs_client.add_group_permissions(
            "sidney",
            [Permission.cancel_job, Permission.modify_subtraction],
            "app",
            "virtool",
        )

        assert await abs_client.list_permissions("ryanf", "app", "virtool") == snapshot


async def test_add_user_permissions(delete_store, spawn_client, snapshot):
    client = await spawn_client(authorize=True)

    abs_client = client.app["auth"]

    await abs_client.add_user_permissions(
        "ryanf",
        [Permission.cancel_job],
        "app",
        "virtool",
    )

    assert await abs_client.list_permissions("ryanf", "app", "virtool") == snapshot


@pytest.mark.parametrize("db", ["mongo", "open_fga"])
async def test_remove_group_permissions(mongo, delete_store, db, setup_update_group, snapshot):
    client, group = setup_update_group

    abs_client = client.app["auth"]

    if db == "mongo":
        await mongo.groups.find_one_and_update(
            {"_id": group.id},
            {"$set": {"permissions.cancel_job": True, "permissions.create_ref": True}},
        )

        await abs_client.remove_group_permissions(group.id, [Permission.cancel_job], "app", "virtool")

        assert (
                await client.db.users.find({}, ["groups", "permissions"]).to_list(None)
                == snapshot
        )

    else:
        await write_tuple(
            abs_client.open_fga, "user", "ryanf", "member", "group", "sidney"
        )

        await write_tuple(
            abs_client.open_fga,
            "group",
            "sidney#member",
            "cancel_job",
            "app",
            "virtool",
        )

        await write_tuple(
            abs_client.open_fga,
            "group",
            "sidney#member",
            "create_ref",
            "app",
            "virtool",
        )

        await abs_client.remove_group_permissions("sidney", [Permission.cancel_job], "app", "virtool")

        assert await abs_client.list_permissions("ryanf", "app", "virtool") == snapshot


async def test_remove_user_permissions(delete_store, spawn_client, snapshot):
    client = await spawn_client(authorize=True)

    abs_client = client.app["auth"]

    await write_tuple(
        abs_client.open_fga,
        "user",
        "ryanf",
        Permission.cancel_job,
        "app",
        "virtool",
    )

    await write_tuple(
        abs_client.open_fga,
        "user",
        "ryanf",
        Permission.modify_subtraction,
        "app",
        "virtool",
    )

    await abs_client.remove_user_permissions("ryanf", [Permission.cancel_job], "app", "virtool")

    assert await abs_client.list_permissions("ryanf", "app", "virtool") == snapshot
