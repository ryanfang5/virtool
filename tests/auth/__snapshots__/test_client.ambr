# name: test_add_group_permissions[uvloop-mongo]
  <class 'list'> [
    <class 'dict'> {
      '_id': 'test',
      'groups': <class 'list'> [
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '7cf872dc',
      'groups': <class 'list'> [
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '3cbb22cc',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': True,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': True,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': 'da42dca3',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': True,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': True,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '09d01419',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': True,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': True,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
  ]
---
# name: test_add_group_permissions[uvloop-open_fga]
  <class 'dict'> {
    'cancel_job': True,
    'create_ref': False,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': True,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_add_user_permissions[uvloop]
  <class 'dict'> {
    'cancel_job': True,
    'create_ref': False,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': False,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_list[uvloop-mongo]
  <class 'dict'> {
    'cancel_job': True,
    'create_ref': False,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': False,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_list[uvloop-open_fga]
  <class 'dict'> {
    'cancel_job': True,
    'create_ref': True,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': False,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_remove_group_permissions[uvloop-mongo]
  <class 'list'> [
    <class 'dict'> {
      '_id': 'test',
      'groups': <class 'list'> [
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '7cf872dc',
      'groups': <class 'list'> [
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '3cbb22cc',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': True,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': 'da42dca3',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': True,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    <class 'dict'> {
      '_id': '09d01419',
      'groups': <class 'list'> [
        'bf1b993c',
      ],
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': True,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
  ]
---
# name: test_remove_group_permissions[uvloop-open_fga]
  <class 'dict'> {
    'cancel_job': False,
    'create_ref': True,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': False,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_remove_user_permissions[uvloop]
  <class 'dict'> {
    'cancel_job': False,
    'create_ref': False,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': True,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
